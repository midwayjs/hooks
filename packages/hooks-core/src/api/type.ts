import type { Merge } from 'type-fest'
import { AsyncFunction } from '../types/common'

export type ArrayToObject<T, R = {}> = T extends [infer First, ...infer Rest]
  ? First extends PromiseLike<infer PromiseValue>
    ? PromiseValue
    : First extends object
    ? Merge<First, ArrayToObject<Rest, R>>
    : ArrayToObject<Rest, R>
  : R

export type ApiRunner<
  Input extends object | void,
  Handler extends AsyncFunction
> = (
  ...args: Input extends void
    ? Parameters<Handler>
    : [...args: Parameters<Handler>, input: Input]
) => ReturnType<Handler>

export enum OperatorType {
  Trigger = 'Trigger',
  Middleware = 'Middleware',
}

export type MetadataHelper = {
  setMetadata: <T = any>(key: any, value: T) => void
  getMetadata: <T = any>(key: any) => T
}

export type ExecuteHelper = {
  result?: any
  getInputArguments?: () => any[]
}

export type Operator<Input> = {
  name: string
  type?: Input
  input?: boolean
  metadata?: (helper: MetadataHelper) => void
  execute?: (helper: ExecuteHelper, next: () => Promise<any>) => Promise<void>
}

export type ExtractInputType<T> = {
  [key in keyof T]: T[key] extends Operator<any> ? T[key]['type'] : void
}

export type BaseTrigger = {
  type: string
  requestClient?: {
    fetcher: string
    client: string
  }
  [key: string]: any
}
