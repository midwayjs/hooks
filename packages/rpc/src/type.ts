import { HttpRequestOptions, RawRequestOptions } from '@midwayjs/hooks-core'

export type RequestContext<T, R = any> = {
  req: T
  res: R
}

export type HttpContext = RequestContext<HttpRequestOptions>

export type NextFunction = () => Promise<any>

export type Middleware<T = any> = (ctx: T, next: NextFunction) => void

export type SetupOptions = {
  fetcher?: Fetcher
  middleware?: Middleware[]
  // request config
  baseURL?: string
  withCredentials?: boolean
}

export type Fetcher = (
  req: HttpRequestOptions,
  options: SetupOptions
) => Promise<any>
