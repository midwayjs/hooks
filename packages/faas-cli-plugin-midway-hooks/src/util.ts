import createDebug from 'debug'
import { REGISTER_INSTANCE } from 'ts-node'

export const setupTsnode = () => {
  if (!process[REGISTER_INSTANCE]) {
    // During blitz interal dev, oclif automaticaly sets up ts-node so we have to check
    require('ts-node').register({
      compilerOptions: { module: 'commonjs' },
      transpileOnly: true,
    })
  }
  require('tsconfig-paths/register')
}

export const debug = createDebug('hooks: next-faas-cli-plugin')

export const argsPath = {
  http: 'ctx.request.body.args',
}
