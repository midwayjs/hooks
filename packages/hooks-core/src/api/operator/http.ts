import { BaseTrigger, MetadataHelper, Operator, OperatorType } from '../type'

export const HttpTriggerType = 'HTTP'

export enum HttpMetadata {
  METHOD = 'Http_Method',
  QUERY = 'Http_Query',
  PARAMS = 'Http_Params',
  HEADERS = 'Http_Headers',
  RESPONSE = 'Http_Response',
}

export enum ResponseMetaType {
  CODE = 'Http_Response_Code',
  HEADER = 'Http_Response_Header',
  CONTENT_TYPE = 'Http_Response_ContentType',
  REDIRECT = 'Http_Response_Redirect',
}

export enum HttpMethod {
  GET = 'GET',
  POST = 'POST',
  PUT = 'PUT',
  DELETE = 'DELETE',
  PATCH = 'PATCH',
  HEAD = 'HEAD',
  OPTIONS = 'OPTIONS',
  ALL = 'ALL',
}

export interface HttpTrigger extends BaseTrigger {
  type: typeof HttpTriggerType
  method: HttpMethod
  path?: string
}

export type HttpInputMetadata = {
  query?: Record<string, string>
  headers?: Record<string, string>
  params?: Record<string, string>
}

function createHttpMethodOperator(method: HttpMethod) {
  return (path?: string) => {
    return {
      name: method,
      metadata({ setMetadata }) {
        setMetadata<HttpTrigger>(OperatorType.Trigger, {
          type: HttpTriggerType,
          method,
          path,
          requestClient: {
            fetcher: 'http',
            client: '@midwayjs/rpc',
          },
        })
      },
    } as Operator<void>
  }
}

// HTTP Method
export const All = createHttpMethodOperator(HttpMethod.ALL)
export const Get = createHttpMethodOperator(HttpMethod.GET)
export const Post = createHttpMethodOperator(HttpMethod.POST)
export const Delete = createHttpMethodOperator(HttpMethod.DELETE)
export const Put = createHttpMethodOperator(HttpMethod.PUT)
export const Patch = createHttpMethodOperator(HttpMethod.PATCH)
export const Head = createHttpMethodOperator(HttpMethod.HEAD)
export const Options = createHttpMethodOperator(HttpMethod.OPTIONS)

// HTTP Helper
export function Query<T extends Record<string, string>>(): Operator<{
  query: T
}> {
  return {
    name: HttpMetadata.QUERY,
    input: true,
    metadata({ setMetadata }) {
      setMetadata(HttpMetadata.QUERY, true)
    },
  }
}

export function Params<T extends Record<string, string>>(): Operator<{
  params: T
}> {
  return {
    name: HttpMetadata.PARAMS,
    input: true,
    metadata({ setMetadata }) {
      setMetadata(HttpMetadata.PARAMS, true)
    },
  }
}

export function Headers<T extends Record<string, string>>(): Operator<{
  headers: T
}> {
  return {
    name: HttpMetadata.HEADERS,
    input: true,
    metadata({ setMetadata }) {
      setMetadata(HttpMetadata.HEADERS, true)
    },
  }
}

export type ResponseMetaData = {
  type: ResponseMetaType
  code?: number
  header?: {
    key: string
    value: string
  }
  url?: string
  contentType?: string
}

export function HttpCode(code: number): Operator<void> {
  return {
    name: 'HttpCode',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetaType.CODE, { code })
    },
  }
}

export function SetHeader(key: string, value: string): Operator<void> {
  return {
    name: 'SetHeader',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetaType.HEADER, {
        header: {
          key,
          value,
        },
      })
    },
  }
}

export function Redirect(url: string, code?: number): Operator<void> {
  return {
    name: 'Redirect',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetaType.REDIRECT, {
        url,
        code,
      })
    },
  }
}

export function ContentType(contentType: string): Operator<void> {
  return {
    name: 'ContentType',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetaType.CONTENT_TYPE, {
        contentType,
      })
    },
  }
}

function setResponseMetaData(
  helper: MetadataHelper,
  type: ResponseMetaType,
  value: Partial<ResponseMetaData>
) {
  const responseMetaData =
    helper.getMetadata<ResponseMetaData[]>(HttpMetadata.RESPONSE) || []

  helper.setMetadata(HttpMetadata.RESPONSE, [
    ...responseMetaData,
    {
      type,
      ...value,
    },
  ])
}
