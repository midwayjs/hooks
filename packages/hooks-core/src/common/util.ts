import parseArgs from 'fn-args'
import debug from 'debug'
import isClass from 'is-class'

export function isHooksMiddleware(fn: (...args: any[]) => any) {
  return typeof fn === 'function' && parseArgs(fn).length === 1 && !isClass(fn)
}

export function extractMetadata(target: any) {
  const metadata: any = {}
  const metaKeys = Reflect.getMetadataKeys(target)
  for (const key of metaKeys) {
    metadata[key] = Reflect.getMetadata(key, target)
  }
  return metadata
}

export function createDebug(namespace: string) {
  return debug(namespace)
}
