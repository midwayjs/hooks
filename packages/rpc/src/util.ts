import type {
  ApiRoute,
  HttpTrigger,
  RequestArgs,
  RequestRoute,
} from '@midwayjs/hooks-core'

type TranspileApiFunction = {
  (...args: any[]): Promise<any>
  route?: ApiRoute<HttpTrigger>
}

export function parseRequestArgs<T, I = void>(requestArgs: RequestArgs<T, I>) {
  const route = requestArgs[requestArgs.length - 1] as RequestRoute<T>
  if (route.useInputMetadata) {
    return {
      trigger: route.trigger,
      metadata: requestArgs[requestArgs.length - 2] as I,
      args: requestArgs.slice(0, requestArgs.length - 2),
    }
  }

  return {
    trigger: route.trigger,
    metadata: null,
    args: requestArgs.slice(0, requestArgs.length - 1),
  }
}

export function args(...inputs: any[]) {
  if (!Array.isArray(inputs)) {
    throw new Error('args must be an array')
  }

  return { args: inputs }
}

export function getHttpTrigger(fn: TranspileApiFunction) {
  return fn.route.trigger
}
