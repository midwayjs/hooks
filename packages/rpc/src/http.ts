import type { FileRecord, HttpRequestOptions } from '@midwayjs/hooks-core'
import { args } from './util'
import { createClient, RequestOptionsCreator } from './client'
import { HttpContext, Middleware, SetupOptions } from './type'
import axios, { Options } from 'redaxios'
import fetch from 'isomorphic-unfetch'

export const client: SetupOptions = {
  baseURL: '',
  async fetcher(req: HttpRequestOptions, options: SetupOptions) {
    const res = await axios({
      method: req.method as Options['method'],
      url: req.url,
      data: req.data,
      params: req.query,
      headers: req.headers,
      baseURL: options.baseURL,
      withCredentials: options.withCredentials,
      fetch,
    })
    return res.data
  },
  middleware: [],
  withCredentials: false,
}

export function setupHttpClient(options: SetupOptions) {
  Object.assign(client, options)
}

export const creator: RequestOptionsCreator<HttpRequestOptions> = (req) => {
  const { trigger, metadata, args: inputs } = req

  return {
    method: trigger.method,
    url:
      typeof metadata?.params === 'object'
        ? format(trigger.path, metadata.params)
        : trigger.path,
    data: inputs.length > 0 ? args(...inputs) : null,

    query: metadata?.query,
    headers: {
      accept: 'application/json',
      ...metadata?.headers,
    },
    files: metadata?.files,
  }
}

const request: Middleware<HttpContext> = async (ctx, next) => {
  ctx.res = await client.fetcher(ctx.req, client)
  return next()
}

const formdata: Middleware<HttpContext> = async (ctx, next) => {
  const files: FileRecord = ctx.req.files
  if (!files) {
    throw new Error('no files')
  }

  const formdata = new FormData()
  for (const [key, value] of Object.entries(files)) {
    if (value instanceof FileList) {
      for (const file of value) {
        formdata.append(key, file)
      }
    } else {
      formdata.append(key, value)
    }
  }

  ctx.req.data = formdata

  return next()
}

export const http = createClient<HttpContext>(creator, () => [
  ...client.middleware,
  request,
])

export const upload = createClient<HttpContext>(creator, () => [
  ...client.middleware,
  formdata,
  request,
])

export function format(url: string, params: Record<string, string>) {
  let result = url
  for (const key in params) {
    result = result.replace(`:${key}`, params[key])
  }
  return result
}
