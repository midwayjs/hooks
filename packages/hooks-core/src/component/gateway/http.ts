import { existsSync } from 'fs'
import staticCache from 'koa-static-cache'
import {
  HTTPRoute,
  ServerRoute,
} from 'packages/hooks-core/dist/midwayjs-hooks-core.cjs'
import { __decorate } from 'tslib'
import { join } from 'upath'

import {
  IMidwayApplication,
  IMidwayContainer,
  IMidwayContext,
} from '@midwayjs/core'
import { All, Controller, Inject, Provide } from '@midwayjs/decorator'

import { superjson } from '../../lib'
import { useContext } from '../../runtime'
import { ApiFunction } from '../../types/common'
import {
  ComponentConfig,
  CreateApiParam,
  HooksGatewayAdapter,
} from '../../types/gateway'
import { isDevelopment, useHooksMiddleware } from '../../util'

export class HTTPGateway implements HooksGatewayAdapter {
  config: ComponentConfig
  container: IMidwayContainer
  app: IMidwayApplication<IMidwayContext>

  constructor(config: ComponentConfig) {
    this.config = config
  }

  is(route: ServerRoute<HTTPRoute>) {
    return !!route?.basePath
  }

  createApi(param: CreateApiParam) {
    const { id, fn, httpPath } = param

    // 设置中间件
    fn.middleware.unshift(...this.getMiddleware())

    // Source: https://www.typescriptlang.org/play?noImplicitAny=false&strictNullChecks=false&strictFunctionTypes=false&strictPropertyInitialization=false&strictBindCallApply=false&noImplicitThis=false&noImplicitReturns=false&alwaysStrict=false&importHelpers=true&emitDecoratorMetadata=false&ts=4.1.5#code/JYWwDg9gTgLgBAbzgSQHYCsCmBjGAaOAYQlRiggBsLMoCBBKggBXIDdgATTOAXzgDNyIOAHIAAiE4B3AIYBPdAGcA9F2zQZMaCIBQO9akXx+qOAF44ACgB0tmVADmigFxwZqOQG0AugEpzAHyIPHpiLBDsXJa+OmLEpORUNJYiyiIx2BQyiopwAGIArqi4wCTxMDLAqDSIOnD1cGJoWLjRdQ24AB6u7nJ6DY0MFCkiBEiSHBzUslCYrj68MQPZcsVwABbuUzXRtQP11PD2TuZwMOvAitZd1rMAjgWYRtYARhAcctbHuQA+P3A+dr7OCzGAFKCmGSyYDGVA2OyORRLBohEJAA
    let FunctionContainer = class FunctionContainer {
      ctx: any
      async handler() {
        let args = this.ctx.request?.body?.args || []
        return await fn(...args)
      }
    }
    __decorate([Inject()], FunctionContainer.prototype, 'ctx', void 0)
    __decorate(
      [All(httpPath, { middleware: fn.middleware })],
      FunctionContainer.prototype,
      'handler',
      null
    )
    FunctionContainer = __decorate(
      [Provide(id), Controller('/')],
      FunctionContainer
    )

    this.container.bind(id, FunctionContainer)
  }

  getMiddleware() {
    const mws = [this.handleError]

    const enableSuperjson = this.config.internal.superjson
    if (enableSuperjson) {
      mws.push(this.deserializeSuperjson)
    }

    return mws.map(useHooksMiddleware)
  }

  async deserializeSuperjson(next: any) {
    await next()
    const ctx = useContext()
    if (ctx.type.includes('application/json')) {
      ctx.body = superjson.serialize(ctx.body)
    }
  }

  handleError = async (next: any) => {
    try {
      await next()
    } catch (error) {
      const ctx = useContext()

      if (this.config.internal.superjson) {
        ctx.status = 500
        ctx.body = superjson.serialize(error)
      } else {
        throw error
      }
    }
  }

  afterCreate() {
    if (isDevelopment()) {
      return
    }

    if (!this.isViteProject) {
      return
    }

    const {
      router: { routes },
    } = this.config
    if (routes.has('/') || routes.has('/*')) {
      return
    }

    const baseDir = this.container.get('baseDir')
    const mw = staticCache({
      dir: join(baseDir, '..', this.config.internal.build.viteOutDir),
      dynamic: true,
      alias: {
        '/': 'index.html',
        /**
         * Add alias for windows, '/' -> '\\'
         * https://github.com/koajs/static-cache/blob/master/index.js#L45
         */
        '\\': 'index.html',
      },
      buffer: true,
      gzip: true,
    })

    const fn: ApiFunction = async () => {}
    fn.middleware = [mw]

    this.createApi({
      id: 'hooks-host',
      httpPath: '/*',
      fn,
    })
  }

  get isViteProject() {
    return ['vite.config.ts', 'vite.config.js'].some((config) =>
      existsSync(join(this.config.root, config))
    )
  }
}
