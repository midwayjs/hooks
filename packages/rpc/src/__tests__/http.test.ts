import {
  HttpInputMetadata,
  HttpTrigger,
  RequestRoute,
} from '@midwayjs/hooks-core'
import { http, Middleware, setupHttpClient } from '..'

import { creator, format } from '../http'
import { HttpContext } from '../type'
import { parseRequestArgs } from '../util'

it('format', () => {
  expect(format('/foo/:bar', { bar: 'baz' })).toBe('/foo/baz')
  expect(format('/:foo/:bar', { foo: 'baz', bar: 'qux' })).toBe('/baz/qux')
  expect(format('/:foo/:bar', { foo: 'baz', bar: 'qux', qux: 'quux' })).toBe(
    '/baz/qux'
  )
})

it('parseRequestArgs & creator', () => {
  {
    const rawOptions = parseRequestArgs([
      'a',
      'b',
      {
        params: {
          bar: 'baz',
        },
      } as HttpInputMetadata,
      {
        trigger: {
          path: '/foo/:bar',
          method: 'GET',
        },
        useInputMetadata: true,
      } as RequestRoute<HttpTrigger>,
    ])

    expect(rawOptions).toMatchSnapshot()
    expect(creator(rawOptions)).toMatchSnapshot()
  }

  {
    const rawOptionsWithoutArgs = parseRequestArgs([
      {
        trigger: {
          path: '/foo',
          method: 'GET',
        },
        useInputMetadata: false,
      } as RequestRoute<HttpTrigger>,
    ])

    expect(rawOptionsWithoutArgs).toMatchSnapshot()
    expect(creator(rawOptionsWithoutArgs)).toMatchSnapshot()
  }

  {
    const methodAll = parseRequestArgs([
      {
        trigger: {
          path: '/foo',
          method: 'ALL',
        },
        useInputMetadata: false,
      } as RequestRoute<HttpTrigger>,
    ])

    expect(methodAll).toMatchSnapshot()
    expect(creator(methodAll)).toMatchSnapshot()
  }

  {
    const methodAllWithArgs = parseRequestArgs([
      'a',
      'b',
      {
        trigger: {
          path: '/foo',
          method: 'ALL',
        },
        useInputMetadata: false,
      } as RequestRoute<HttpTrigger>,
    ])

    expect(methodAllWithArgs).toMatchSnapshot()
    expect(creator(methodAllWithArgs)).toMatchSnapshot()
  }
})

describe('http', () => {
  test('normal', async () => {
    const fetcher = jest.fn()
    fetcher.mockReturnValue({ isTest: true })
    setupHttpClient({ fetcher })
    const response = await http({
      trigger: {
        path: '/foo',
        method: 'GET',
      },
      useInputMetadata: false,
    } as RequestRoute<HttpTrigger>)

    expect(fetcher).toBeCalledTimes(1)
    expect(fetcher.mock.calls).toMatchSnapshot()
    expect(response).toEqual({ isTest: true })
  })

  test('withMiddleware', async () => {
    const fetcher = jest.fn()
    fetcher.mockReturnValue({ isTest: true })

    const middleware: Middleware<HttpContext> = async (ctx, next) => {
      expect(ctx.req.method).toEqual('GET')
      expect(ctx.req.url).toEqual('/foo')
      await next()
      expect(ctx.res).toEqual({ isTest: true })
      ctx.res = { ...ctx.res, isMiddleware: true }
    }
    setupHttpClient({ fetcher, middleware: [middleware] })

    const response = await http({
      trigger: {
        path: '/foo',
        method: 'GET',
      },
      useInputMetadata: false,
    } as RequestRoute<HttpTrigger>)
    expect(fetcher).toBeCalledTimes(1)
    expect(response).toEqual({ isTest: true, isMiddleware: true })
  })
})
