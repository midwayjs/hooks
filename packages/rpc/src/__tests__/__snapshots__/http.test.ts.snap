// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`http normal 1`] = `
Array [
  Array [
    Object {
      "data": null,
      "files": undefined,
      "headers": Object {
        "accept": "application/json",
      },
      "method": "GET",
      "query": undefined,
      "url": "/foo",
    },
    Object {
      "baseURL": "",
      "fetcher": [MockFunction] {
        "calls": [Circular],
        "results": Array [
          Object {
            "type": "return",
            "value": Object {
              "isTest": true,
            },
          },
        ],
      },
      "middleware": Array [],
      "withCredentials": false,
    },
  ],
]
`;

exports[`parseRequestArgs & creator 1`] = `
Object {
  "args": Array [
    "a",
    "b",
  ],
  "metadata": Object {
    "params": Object {
      "bar": "baz",
    },
  },
  "trigger": Object {
    "method": "GET",
    "path": "/foo/:bar",
  },
}
`;

exports[`parseRequestArgs & creator 2`] = `
Object {
  "data": Object {
    "args": Array [
      "a",
      "b",
    ],
  },
  "files": undefined,
  "headers": Object {
    "accept": "application/json",
  },
  "method": "GET",
  "query": undefined,
  "url": "/foo/baz",
}
`;

exports[`parseRequestArgs & creator 3`] = `
Object {
  "args": Array [],
  "metadata": null,
  "trigger": Object {
    "method": "GET",
    "path": "/foo",
  },
}
`;

exports[`parseRequestArgs & creator 4`] = `
Object {
  "data": null,
  "files": undefined,
  "headers": Object {
    "accept": "application/json",
  },
  "method": "GET",
  "query": undefined,
  "url": "/foo",
}
`;

exports[`parseRequestArgs & creator 5`] = `
Object {
  "args": Array [],
  "metadata": null,
  "trigger": Object {
    "method": "ALL",
    "path": "/foo",
  },
}
`;

exports[`parseRequestArgs & creator 6`] = `
Object {
  "data": null,
  "files": undefined,
  "headers": Object {
    "accept": "application/json",
  },
  "method": "GET",
  "query": undefined,
  "url": "/foo",
}
`;

exports[`parseRequestArgs & creator 7`] = `
Object {
  "args": Array [
    "a",
    "b",
  ],
  "metadata": null,
  "trigger": Object {
    "method": "ALL",
    "path": "/foo",
  },
}
`;

exports[`parseRequestArgs & creator 8`] = `
Object {
  "data": Object {
    "args": Array [
      "a",
      "b",
    ],
  },
  "files": undefined,
  "headers": Object {
    "accept": "application/json",
  },
  "method": "POST",
  "query": undefined,
  "url": "/foo",
}
`;
